src/core/enum_type.hpp

```diff
@@ -75,4 +75,56 @@ debug_inline constexpr void ToggleFlag(T &x, const T y)
 	}
 }
 
+template <typename Tenum, typename Tstorage>
+class EnumBitSet {
+public:
+	constexpr EnumBitSet() : data(0) {}
+	constexpr EnumBitSet(Tenum other) : data(to_underlying(other)) {}
+	constexpr EnumBitSet(Tstorage data) : data(data) {}
+
+	constexpr EnumBitSet(const EnumBitSet &) = default;
+
+	constexpr auto operator <=>(const EnumBitSet &) const noexcept = default;
+
+	inline constexpr EnumBitSet &set(Tenum value)
+	{
+		this->data |= (1U << to_underlying(value));
+		return *this;
+	}
+
+	inline constexpr EnumBitSet &clear(Tenum value)
+	{
+		this->data &= ~(1U << to_underlying(value));
+		return *this;
+	}
+
+	inline constexpr EnumBitSet &toggle(Tenum value)
+	{
+		if (this->test(value)) {
+			return this->clear(value);
+		} else {
+			return this->set(value);
+		}
+	}
+
+	inline constexpr bool test(Tenum value) const
+	{
+		return (this->data & (1U << to_underlying(value))) != 0;
+	}
+
+	inline constexpr EnumBitSet operator |(Tenum value) const
```

Comment:
Maybe a bit of documentation here that it is an atypical usage of `|`, i.e. that not the literal value is taken but that it's made the appropriate bitmask first.

Alternatively, won't Flags(Flag::First).set(Flag::Second).set(Flag::Third) work equally well and be less confusing?

Reply:
I've changed this now to a more conventional `inline constexpr EnumBitSet operator |(const EnumBitSet &other) const` which allows combining two EnumBitSets. Due to implicit conversion the original still works, perhaps a bit magic still.